modules/dnn/src/ocl4dnn/src/math_functions.cpp

```diff
@@ -0,0 +1,674 @@
+//  copy or use the software.
+//
+//
+//                           License Agreement
+//                For Open Source Computer Vision Library
+//
+// Copyright (C) 2017, Intel Corporation, all rights reserved.
+// Copyright (c) 2016-2017 Fabian David Tschopp, all rights reserved.
+// Third party copyrights are property of their respective owners.
+//
+// Redistribution and use in source and binary forms, with or without modification,
+// are permitted provided that the following conditions are met:
+//
+//   * Redistribution's of source code must retain the above copyright notice,
+//     this list of conditions and the following disclaimer.
+//
+//   * Redistribution's in binary form must reproduce the above copyright notice,
+//     this list of conditions and the following disclaimer in the documentation
+//     and/or other materials provided with the distribution.
+//
+//   * The name of the copyright holders may not be used to endorse or promote products
+//     derived from this software without specific prior written permission.
+//
+// This software is provided by the copyright holders and contributors "as is" and
+// any express or implied warranties, including, but not limited to, the implied
+// warranties of merchantability and fitness for a particular purpose are disclaimed.
+// In no event shall the Intel Corporation or contributors be liable for any direct,
+// indirect, incidental, special, exemplary, or consequential damages
+// (including, but not limited to, procurement of substitute goods or services;
+// loss of use, data, or profits; or business interruption) however caused
+// and on any theory of liability, whether in contract, strict liability,
+// or tort (including negligence or otherwise) arising in any way out of
+// the use of this software, even if advised of the possibility of such damage.
+//
+//M*/
+
+#include "../../precomp.hpp"
+#include "common.hpp"
+
+#ifdef HAVE_OPENCL
+#include "math_functions.hpp"
+#include <vector>
+#include "opencl_kernels_dnn.hpp"
+
+using namespace cv;
+
+struct gemm_callback_arg {
+    std::vector<cl_event> evs;
+    std::vector<cl_mem> imgs;
+};
+
+static void CL_CALLBACK gemmCallback(cl_event event,
+                                     cl_int event_command_exec_status,
+                                     void *user_data)
+{
+    struct gemm_callback_arg *arg = (struct gemm_callback_arg *) user_data;
+    for(int i = 0; i < arg->evs.size(); i++)
+    {
+        clReleaseEvent(arg->evs[i]);
+    }
+
+    for(int i = 0; i < arg->imgs.size(); i++)
+    {
+        clReleaseMemObject(arg->imgs[i]);
+    }
+    delete arg;
+}
+
+// Create and copy buffer to image for GEMM's matrix A and B.
+// Will return image to caller if the input image is NULL. Otherwise,
+// will use the image directly. It's caller's responsibility to
+// release the created image.
+template<typename Dtype>
+void ocl4dnnGEMMCopyBufferToImage(int32_t ctx_id,
```

Comment:
ocl4dnnGEMMCopyBufferToImage isn't just copy the umat data as is, it may transpose the matrix and add extra padding for purpose.  So cv::ocl::Image2D isn't suitable in this case.